packages/builders/src/components/Assertions.ts

```diff
@@ -1,127 +1,160 @@
-import { s } from '@sapphire/shapeshift';
-import { ButtonStyle, ChannelType, type APIMessageComponentEmoji } from 'discord-api-types/v10';
-import { isValidationEnabled } from '../util/validation.js';
-import { StringSelectMenuOptionBuilder } from './selectMenu/StringSelectMenuOption.js';
-
-export const customIdValidator = s
-	.string()
-	.lengthGreaterThanOrEqual(1)
-	.lengthLessThanOrEqual(100)
-	.setValidationEnabled(isValidationEnabled);
-
-export const emojiValidator = s
+import { ButtonStyle, ChannelType, ComponentType, SelectMenuDefaultValueType } from 'discord-api-types/v10';
+import { z } from 'zod';
+import { customIdPredicate, refineURLPredicate } from '../Assertions.js';
+
+const labelPredicate = z.string().min(1).max(80);
+
+export const emojiPredicate = z
 	.object({
-		id: s.string(),
-		name: s.string(),
-		animated: s.boolean(),
+		id: z.string().optional(),
+		name: z.string().optional(),
+		animated: z.boolean().optional(),
 	})
```

Comment:
We should refine this to ensure the user passed either `id` or `name`

Something like this could work
```ts
  .refine((data) => data.id !== undefined || data.name !== undefined, {
    message: "Either 'id' or 'name' must be provided",
    path: ['id', 'name'],
  });
```

Reply:
I don't think this is how you use path, I think `path: ['id', 'name']` implies `<X>.id.name`

https://zod.dev/ERROR_HANDLING?id=zodissue